### ISSUE TITLE
Latency of Teensy 3.2 keyboard

### ISSUE THREAD

**Issue by vatthaphon** [sappelhoff/usb-to-ttl#6]
Dear Stefan,

In Windows 10 without any program opened, I tried to measure the input lag (the latency due to the USB) that is the latency of Teensy 3.2 keyboard in Table 1 of your "In COM we trust" paper, if I understood your paper correctly. I used pyserial in python. I got a clear bimodal Gaussian distribution: one has a mean around 1 ms. and another around 1.5 ms. From your paper, you observed a unimodal Gaussian distribution. I suspect a reason that I got the bimodal distribution is due to Windows 10 I/O management. Could you please suggest me if there is any required windows 10 setting to get the unimodal distribution? Thank you in advance.

Best,
Atthaphon

**Comment by sappelhoff** [comment]
Hi @vatthaphon could you provide some more details what you measured and how? (including the code, maybe?)

also cc @tstenner

**Comment by vatthaphon** [comment]
Hi Stefan,

Thank you for your quick response :)

In summary, after some tweaking described below, the RTT of a USB connection distributes more like a gaussian,  c.f. the box plot given in https://www.dropbox.com/s/scb83x7b5wy5kex/RTT.png?dl=0 . RTT is calculated according to the clock synchronization algorithm, c.f. https://en.wikipedia.org/wiki/Network_Time_Protocol . So the USB latency, defined as a time for sending a byte from my MCU to PC, is around 0.5 ms. 

My hardware setup:
- ESP-WROOM-32
- An USB cable connecting to USB 2.0 port
- A PC

My software setup:
- Windows 10
- ESP-WROOM-32 deals with an arrival of a byte using an interrupt service routine. In ISR, as soon as ISR is invoked, a counting byte is sent back to PC.
- In PC, I used a Python with pyserial. I checked if the counting byte is as I expected, so that no garbage is received.

Best,
Atthaphon

**Comment by sappelhoff** [comment]
That's good to hear, the plot looks good to me! I think that what you are measuring is slightly different from what we measured in the paper though (see Fig 1 therein). If I got you correctly, then you are measuring a "roundtrip" latency: From a PC, send a signal to a device at t1, then make the device send the signal back after receiving it, and once the PC receives that signal at t2, compute the delta: t2-t1. So I guess some slight deviation in values (to the values we report in the paper) is to be expected.
